### PR TITLE
Introducing learning rate decay + option for reusing embeddings for Autoregressive decoders

### LINKED ISSUE #631: Deprecate config.utils
`neuralmonkey.config.utils` should be deprecated or removed because we can directly use `tf.train.AdamOptimizer`, `tf.Variable` etc.

### PR DESCRIPTION
**Pull request by varisd**
- Reworked optimizers functions
- Introduced lr_decay function that works with optimizers
- AutoregressiveDecoder derived classes can now reuse embeddings from encoder
- Transformer Decoder decoding matrix replaced with transposed embedding matrix, thus further reducing number of model parameters.

Closes #631

### CODE REVIEW

File: tests/bpe.ini [hunk 1]
```diff
@@ -51,15 +51,22 @@ preprocessors=[("source", "source_bpe", <bpe_preprocess>)]
...
-
 [bpe_postprocess]
 class=processors.bpe.BPEPostprocessor
 
+;[bpe_vocabulary]
+;class=vocabulary.from_bpe
+;path="tests/data/merges_100.bpe"
 
+; The vocabulary.from_bpe method is incorrect.
+; Use vocabulary.from_dataset of vocabulary.from_file instead.
```
Comment (jindrahelcl) [inline tests/bpe.ini]: to zahoď

Comment (varisd) [reply tests/bpe.ini]: Ok, ale nekde bychom to meli mit poznacene, dokud tu metodu uplne nevyhodime.

Comment (jindrahelcl) [reply tests/bpe.ini]: Tak rovnou udělej PR kterej do ní dá raise NotImplementedError s odkazem na tu issue na githubu

File: neuralmonkey/trainers/optimizers.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
+"""Collection of helper functions providing optimizers for trainers."""
```
Comment (jlibovicky) [inline neuralmonkey/trainers/optimizers.py]: Jak už tady někde psal @cifkao,  tyhlety helpery vůbec nepotřebujeme, kontruktory optimizerů sde dají volat rovnou z konfiguráku. Ten learning rate decay be dal pak pořešit tím, že místo `learning_rate * lr_decay(global_step)`, které je v tomhle souboru třikrát by funkce, by decay vracel tensor, co by byl výsledek toho násobení. V případě konstrantního learning ratu, by se napsal do konfiguráku float.

File: neuralmonkey/trainers/lr_decay.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
+"""Collection of learning rate decay strategies."""
```
Comment (jlibovicky) [inline neuralmonkey/trainers/lr_decay.py]: Zkusil bych se zamyslet nad tím, jestli by to nešlo pojmenovat nějak obecněji - ve světle toho článku, co doporučuje ten learning rate sinusovkovat.

File: neuralmonkey/trainers/generic_trainer.py [hunk 4]
```diff
@@ -39,8 +45,23 @@ def __init__(self, objectives: List[Objective],
...
-            self.optimizer = optimizer or tf.train.AdamOptimizer(1e-4)
+            if global_step is None:
+                global_step = tf.Variable(
+                    0, trainable=False, name="global_step")
+            self.global_step = global_step
+
+            self.optimizer = optimizer_getter(  # type: ignore
+                self.global_step,
+                decay_function)
+            # TODO: avoid accessing private member
```
Comment (jlibovicky) [inline neuralmonkey/trainers/generic_trainer.py]: Tohle TODO bych drze zahodil a neštil bych se dál používat `_lr`, ono se to třeba rozbije s novou verzí tensorflow.

File: neuralmonkey/trainers/generic_trainer.py [hunk 5]
```diff
@@ -39,8 +45,23 @@ def __init__(self, objectives: List[Objective],
...
+            # pylint: disable=protected-access
+            tf.summary.scalar("learning_rate", self.optimizer._lr,
```
Comment (jlibovicky) [inline neuralmonkey/trainers/generic_trainer.py]: Kreslil bych ten, jenom když ten lr bude tensor, když bude float, tak bych na to kašlal. Už tahle jsou ty events fily obrovský, nepotřebujeme tam psát konstantu.

File: neuralmonkey/decoders/decoder.py [hunk 6]
```diff
@@ -58,13 +57,13 @@ def __init__(self,
                  name: str,
                  max_output_len: int,
                  dropout_keep_prob: float = 1.0,
+                 embedding_size: int = None,
```
Comment (jlibovicky) [inline neuralmonkey/decoders/decoder.py]: Nepromítlo se v docstringu.

File: neuralmonkey/config/utils.py [hunk 7]
```diff
@@ -4,11 +4,7 @@
 the configuration file because calling the functions or the class constructors
 directly would be inconvinent or impossible.
 """
-
 from typing import Callable, TypeVar
-
-import tensorflow as tf
-
 from neuralmonkey.logging import warn
```
Comment (jlibovicky) [inline neuralmonkey/config/utils.py]: To deprecated bych někam přestěhoval a ten soubor smazal. Vždyť to vůbec nesouhlasí s deklarovaným smyslem tohodle modulu v tom docstringu.

File: neuralmonkey/trainers/optimizers.py (hunk 2 above)
Comment (cifkao) [reply neuralmonkey/trainers/optimizers.py]: Souhlasím.

Možná by nebylo od věci to API udělat tak, aby se tam dalo použít např. [`tf.train.exponential_decay`](https://www.tensorflow.org/api_docs/python/tf/train/exponential_decay).

File: neuralmonkey/config/utils.py (hunk 7 above)
Comment (varisd) [reply neuralmonkey/config/utils.py]: Ok, ale to uz je spis na separatni issue.

File: neuralmonkey/trainers/cross_entropy_trainer.py [hunk 8]
```diff
@@ -21,10 +24,15 @@ def xent_objective(decoder, weight=None) -> Objective:
...
+    def __init__(self,
+                 decoders: List[Any],
                  decoder_weights: List[ObjectiveWeight] = None,
-                 l1_weight=0., l2_weight=0.,
-                 clip_norm=False, optimizer=None, global_step=None,
+                 l1_weight: float = 0.,
+                 l2_weight: float = 0.,
+                 clip_norm: float = None,
+                 optimizer_getter: OptimizerGetter = adam_optimizer(),
+                 global_step: tf.Tensor = None,
```
Comment (cifkao) [inline neuralmonkey/trainers/cross_entropy_trainer.py]: Já bych asi ten `global_step` všude nepředával a použil [`tf.train.get_or_create_global_step()`](https://www.tensorflow.org/api_docs/python/tf/train/get_or_create_global_step). To se v případě potřeby dá zavolat i z konfiguráku, např.:

```ini
[lr_decay]
class=tf.train.exponential_decay
learning_rate=0.1
decay_steps=100000
decay_rate=0.96
global_step=<global_step>

[global_step]
class=tf.train.get_or_create_global_step
```

A námi definované decay funkce ten `global_step` mít jako povinný parametr nemusejí a zjistí si ho samy.

File: neuralmonkey/trainers/lr_decay.py (hunk 3 above)
Comment (varisd) [reply neuralmonkey/trainers/lr_decay.py]: Tak obecne muzeme zavest learning_rate_modifier (nebo decay) funkce, ktere budou dostavat lr a step a budou vracet v case upraveny lr.

Comment (varisd) [reply neuralmonkey/trainers/lr_decay.py]: Nejsem proti mit vsechny tyto funkce mit nahazene v jednom zdrojaku (treba ten neuralmonkey.functions). A v budoucnu az by jich bylo hodne, tak by se to treba rozhazelo do vice zdrojaku.

File: neuralmonkey/trainers/cross_entropy_trainer.py (hunk 8 above)
Comment (varisd) [reply neuralmonkey/trainers/cross_entropy_trainer.py]: Supr, tohle jsem neznal :)

File: neuralmonkey/decoders/decoder.py (hunk 6 above)
Comment (varisd) [reply neuralmonkey/decoders/decoder.py]: DONE

File: neuralmonkey/trainers/optimizers.py (hunk 2 above)
Comment (jindrahelcl) [reply neuralmonkey/trainers/optimizers.py]: yes. cokoliv co bude rozumnější. Už jsem předesílal, že to, jak je to udělaný teď, se mi vůbec nelíbí

**Comment by varisd** [comment]
ping

File: neuralmonkey/functions.py [hunk 9]
```diff
@@ -50,3 +54,29 @@ def piecewise_function(param, values, changepoints, name=None,
...
+    inv_sq_dim = 1 / math.sqrt(model_dimension)
+    inv_sq3_warmup_steps = math.pow(warmup_steps, -1.5)
+
+    inv_sq_step = 1 / tf.sqrt(step)
+    warmup = step * inv_sq3_warmup_steps
+
+    return learning_rate * inv_sq_dim * tf.minimum(inv_sq_step, warmup)
+
+
+def variable(initial_value: Any = 0,
```
Comment (cifkao) [inline neuralmonkey/functions.py]: Tohle může pryč, `tf.Variable` se dá použít rovnou...

File: neuralmonkey/train.py [hunk 10]
```diff
@@ -34,8 +35,13 @@ def create_config() -> Configuration:
...
     config.add_argument("test_datasets", required=False, default=[])
-    config.add_argument("logging_period", required=False, default=20)
-    config.add_argument("validation_period", required=False, default=500)
+    # a little "hackish" checking
+    config.add_argument("logging_period",
+                        cond=lambda x: int(re.sub("[^0-9-]", "", str(x))) > 0,
+                        required=False, default=20)
+    config.add_argument("validation_period",
+                        cond=lambda x: int(re.sub("[^0-9-]", "", str(x))) > 0,
+                        required=False, default=500)
```
Comment (cifkao) [inline neuralmonkey/train.py]: K čemu je tahle kontrola dobrá? Pak se to stejně zkontroluje v training loopu. Jestli je nutnost kontroly tady, tak bych zavolal `learning_utils._resolve_period`, ať je aspoň důsledná.

Comment (varisd) [reply neuralmonkey/train.py]: Souhlasim, rad tu kontrolu odsud vyhodim. Prisel s ni Jindra H.

**Comment by jindrahelcl** [comment]
Vyhoď ``config.utils`` úplně, ta deprecated věc se nepoužívá:
Closes #631

**Comment by jindrahelcl** [comment]
@varisd jak to vypada?

**Comment by varisd** [comment]
Pokud nezbobruji testy, tak snad uz hotovo.

File: neuralmonkey/decoders/transformer.py [hunk 11]
```diff
@@ -198,15 +225,25 @@ def layer(self, level: int, inputs: tf.Tensor,
...
+
+        # TODO: Add bias after matmul by embedding_matrix?
 
+        # Reusing input embedding matrix for generating logits
+        # significantly reduces the overall size of the model.
+        # See: https://arxiv.org/pdf/1608.05859.pdf
+        #
         # shape (batch, time, vocab)
-        logits = logits_unbiased + tf.reshape(self.decoding_b, [1, 1, -1])
+        logits = tf.reshape(
```
Comment (jlibovicky) [inline neuralmonkey/decoders/transformer.py]: Že do toho ještě kecám, ale já bych tohle udělal přepínací a odstěhoval to rodiče.

File: neuralmonkey/decoders/transformer.py [hunk 12]
```diff
@@ -198,15 +225,25 @@ def layer(self, level: int, inputs: tf.Tensor,
...
+            tf.matmul(
+                last_layer_states,
+                tf.transpose(self.embedding_matrix)),
+            [last_layer_shape[0], last_layer_shape[1], len(self.vocabulary)])
```
Comment (jlibovicky) [inline neuralmonkey/decoders/transformer.py]: Když se sdílí ta matice, tak tam není bias?

File: neuralmonkey/functions.py [hunk 13]
```diff
@@ -50,3 +53,23 @@ def piecewise_function(param, values, changepoints, name=None,
         predicates = [tf.less(param, x) for x in changepoints]
         return tf.case(list(zip(predicates, lambdas[:-1])), lambdas[-1],
                        name=s_name)
+
+
+def noam_decay(learning_rate: float,
+               model_dimension: int,
+               warmup_steps: int) -> tf.Tensor:
+    """Return decay function as defined in Vaswani et al., 2017, Equation 3.
```
Comment (jlibovicky) [inline neuralmonkey/functions.py]: Link na arxiv, klidně bych tam vylatexoval tu rovnici.

File: tests/bandit.ini [hunk 14]
```diff
@@ -84,11 +84,14 @@ class=trainers.generic_trainer.GenericTrainer
 objectives=[<bandit>]
 l2_weight=1.0e-8
 clip_norm=1.0
-optimizer=<optimizer>
+optimizer=<lazyadam_g>
```
Comment (jlibovicky) [inline tests/bandit.ini]: Proč se jmenuje lazy adam, když není lazy?

File: neuralmonkey/decoders/transformer.py (hunk 12 above)
Comment (varisd) [reply neuralmonkey/decoders/transformer.py]: Ja si myslim, ze by tam nemel byt (jsou to prece jen extra promenne). Podivam se do toho clanku, jestli to tam neni popsane.

Comment (jindrahelcl) [reply neuralmonkey/decoders/transformer.py]: Bias nad slovníkem neuškodí, ne?

Comment (varisd) [reply neuralmonkey/decoders/transformer.py]: Pokud nic neprida, tak to jsou zbytecne promenne navic (viz vyse).

File: neuralmonkey/decoders/autoregressive.py [hunk 15]
```diff
@@ -120,6 +144,8 @@ def batch_size(self) -> tf.Tensor:
     @tensor
     def decoding_w(self) -> tf.Variable:
         with tf.name_scope("output_projection"):
+            if self.tie_embeddings:
+                return tf.transpose(self.embedding_matrix)
```
Comment (jlibovicky) [inline neuralmonkey/decoders/autoregressive.py]: Tady je potřeba kontrolovat, jestli dimenze hidden statů odpovídá dimenzi embeddingů, jinak se to nemůže stát.

File: neuralmonkey/decoders/decoder.py [hunk 16]
```diff
@@ -58,13 +57,14 @@ def __init__(self,
...
+                 embeddings_source: EmbeddedSequence = None,
+                 tie_embeddings: bool = False,
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/decoder.py]: Nepromítlo se v docstringu.

File: neuralmonkey/decoders/autoregressive.py [hunk 17]
```diff
@@ -119,6 +143,14 @@ def batch_size(self) -> tf.Tensor:
 
     @tensor
     def decoding_w(self) -> tf.Variable:
+        if (self.tie_embeddings
+                and self.embedding_size != self.output_dimension):
+            raise ValueError(
+                "`embedding_size must be equal to the output_projection "
+                "size when using the `tie_embeddings` option")
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/autoregressive.py]: tadyta kontrola by měla bejt asi už v konstruktoru. plus nevim co se stane když budeš porovnávat `int` a `tf.Tensor`.

Comment (jindrahelcl) [reply neuralmonkey/decoders/autoregressive.py]: (`self.output_dimension` může bejt `tf.Tensor`)

Comment (jindrahelcl) [reply neuralmonkey/decoders/autoregressive.py]: možná bych tu kontrolu vyhodil

File: neuralmonkey/decoders/autoregressive.py [hunk 18]
```diff
@@ -119,6 +143,14 @@ def batch_size(self) -> tf.Tensor:
...
+
+        if self.tie_embeddings:
+            return tf.transpose(self.embedding_matrix)
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/autoregressive.py]: add empty line after return

Comment (jindrahelcl) [reply neuralmonkey/decoders/autoregressive.py]: a ten transpose komand bych dal do name_scope output_projection.

File: neuralmonkey/decoders/transformer.py [hunk 19]
```diff
@@ -48,10 +49,38 @@ def __init__(self,
                  depth: int,
                  max_output_len: int = None,
                  dropout_keep_prob: float = 1.0,
+                 embedding_size: int = None,
+                 embeddings_source: EmbeddedSequence = None,
+                 tie_embeddings: bool = True,
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/transformer.py]: docstring

File: neuralmonkey/decoders/transformer.py [hunk 20]
```diff
@@ -280,8 +319,11 @@ def body(*args) -> LoopState:
                 # (batch, state_size)
                 output_state = last_layer.temporal_states[:, -1, :]
 
-                logits = tf.matmul(output_state, self.decoding_w)
-                logits += self.decoding_b
+                # See train_logits definition
+                logits_unbiased = tf.matmul(
+                    output_state,
+                    self.decoding_w)
+                logits = logits_unbiased + self.decoding_b
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/transformer.py]: proč je tu tahle změna? logits_unbiased se nikde dál nepoužívaj

File: neuralmonkey/decoders/autoregressive.py (hunk 17 above)
Comment (varisd) [reply neuralmonkey/decoders/autoregressive.py]: V konstruktoru to byt nemuze, protoze AutoregressiveDecoder.output_dimension je abstract_property, ktera je definovana az v potomkovi, ktery se inicializuje az po predkovi.

Comment (varisd) [reply neuralmonkey/decoders/autoregressive.py]: Klidne bych to vyhodil, ale tu kontrolu jsem tam dal na podnet @jlibovicky

File: neuralmonkey/decoders/transformer.py (hunk 20 above)
Comment (varisd) [reply neuralmonkey/decoders/transformer.py]: Bud trochu konzistentni. Presne takto jsi to tam mel ty, i presto, ze jsi potom logits_unbiased nikde dal nepouzival. Ale klidne to prejmenuju.

File: neuralmonkey/decoders/autoregressive.py (hunk 17 above)
Comment (jindrahelcl) [reply neuralmonkey/decoders/autoregressive.py]: No tak jak je to teď je to zkrátka špatně, protože `3 != tf.constant(3)` a `output_dimension` může vracet tenzor.

Comment (varisd) [reply neuralmonkey/decoders/autoregressive.py]: A musi vubec output_dimension byt tf.Tensor? Kdyz se koukam do zdrojaku, tak vzdycky tam je int.

Comment (jindrahelcl) [reply neuralmonkey/decoders/autoregressive.py]: to je potřeba ověřit. Ale záleží to na tom, co použiješ za output projection u dekodéru. některý output projekce to podle mě staticky mít nemusí.

Comment (varisd) [reply neuralmonkey/decoders/autoregressive.py]: Takze:
transformer.output_dimension je int primo z definice
decoder.output_dimension je budto decoder.rnn_size (int) a nebo decoder.output_projection_spec[1] (opet int)

Comment (varisd) [reply neuralmonkey/decoders/autoregressive.py]: Btw, nemel by to nahodou mypy detekovat, pokud by to tak nebylo?

Comment (jindrahelcl) [reply neuralmonkey/decoders/autoregressive.py]: mypy bere `tf.Tensor` jako `Any`, takže ti tady nic nenahlásí

File: neuralmonkey/decoders/transformer.py (hunk 20 above)
Comment (jindrahelcl) [reply neuralmonkey/decoders/transformer.py]: no koukám že se ze dvou čitelnejch řádků stalo pět nečitelnejch a dělaj to samý
